raiden/network/transport/matrix/transport.py

```diff
@@ -1012,12 +1015,46 @@ def _get_room_for_address(
             self.log.error('No valid peer found', peer_address=address_hex)
             return None
 
-        self._address_to_userids[address].update({user.user_id for user in peers})
-
         if self._private_rooms:
             room = self._get_private_room(invitees=peers)
         else:
             room = self._get_public_room(room_name, invitees=peers)
+
+        peer_ids = self._address_to_userids[address]
+        member_ids = {member.user_id for member in room.get_joined_members(force_resync=True)}
+        room_is_empty = not bool(peer_ids & member_ids)
+        if room_is_empty:
+            last_ex: Optional[Exception] = False
+            retry_interval = 0.1
+            self.log.debug(
+                'Waiting for peer to join from invite',
+                peer_address=address_hex,
+            )
+            for _ in range(JOIN_RETRIES):
+                try:
+                    member_ids = {member.user_id for member in room.get_joined_members()}
+                except MatrixRequestError as e:
+                    last_ex = e
+                room_is_empty = not bool(peer_ids & member_ids)
+                if room_is_empty or last_ex:
+                    if last_ex:
+                        raise last_ex
+                    gevent.sleep(retry_interval)
```

Comment:
Even if not going for a separate function, I'd still suggest to use `stop_event.wait` instead of `gevent.sleep` and breaking early in case it returns `True` (was set), so it properly return in case client is stopping, instead of lying around way longer than any Runnable's child is supposed to

Reply:
Done, changed it `_handle_invite` as well.